File: matiasmasso/Mat.NET/DTO/DTOCustomerTarifaDto.cs
using System;
using System.Collections.Generic;

namespace DTO
{
    public class DTOCustomerTarifaDto : DTOBaseGuid
    {
        public DTOBaseGuid CustomerOrChannel { get; set; }
        public string CustomerOrChannelNom { get; set; }
        public DateTime Fch { get; set; }
        public DTOProduct Product { get; set; }
        public decimal Dto { get; set; }
        public Srcs Src { get; set; }
        public string Obs { get; set; }

        public enum Srcs
        {
            notSet,
            canal,
            client
        }

        public new String ToString()
        {
            System.Text.StringBuilder sb = new System.Text.StringBuilder();
            sb.Append("DTOCustomerTarifaDto: ");
            sb.AppendFormat("{0}%", this.Dto);
            sb.AppendFormat(" {dd/MM/yy}", this.Fch);
            string retval = sb.ToString();
            return retval;
        }

        public DTOCustomerTarifaDto() : base()
        {
        }


        public DTOCustomerTarifaDto(Guid oGuid) : base(oGuid)
        {
        }

        public static DTOCustomerTarifaDto Factory(DTOBaseGuid oCustomerOrChannel, string Nom = "")
        {
            DTOCustomerTarifaDto retval = new DTOCustomerTarifaDto();
            {
                retval.CustomerOrChannel = oCustomerOrChannel;
                if (oCustomerOrChannel is DTOContact)
                    retval.Src = Srcs.client;
                else if (oCustomerOrChannel is DTODistributionChannel)
                    retval.Src = Srcs.canal;
                retval.CustomerOrChannelNom = Nom;
                retval.Fch = DTO.GlobalVariables.Now();
            }
            return retval;
        }

        public static string FullNom(DTOCustomerTarifaDto oCustomerDto, DTOLang oLang = null/* TODO Change to default(_) if this is not a reference type */)
        {
            if (oLang == null)
                oLang = DTOApp.Current.Lang;
            string sAllBrands = oLang.Tradueix("todas las marcas", "totes les marques", "all brands");
            string sConcept = (oCustomerDto == null) ? sAllBrands : oCustomerDto.Product.Nom.Esp;
            string retval = string.Format("{0:dd/MM/yy} {1}", oCustomerDto.Fch, sConcept);
            return retval;
        }

        public static decimal ProductDto(List<DTOCustomerTarifaDto> ActiveItems, DTOProduct oProduct)
        {
            decimal retval = 0;

            DTOCustomerTarifaDto oDto = null;
            if (oProduct is DTOProductSku oSku)
            {
                DTOProductCategory oCategory = (DTOProductCategory)oSku.Category;
                DTOProductBrand oBrand = oCategory.Brand;
                oDto = ActiveItems.Find(x => oSku.Equals(x.Product));
                if (oDto == null)
                {
                    oDto = ActiveItems.Find(x => oCategory.Equals(x.Product));
                    if (oDto == null)
                        oDto = ActiveItems.Find(x => oBrand.Equals(x.Product));
                }
            }
            else if (oProduct is DTOProductCategory oCategory)
            {
                DTOProductBrand oBrand = oCategory.Brand;
                oDto = ActiveItems.Find(x => oCategory.Equals(x.Product));
                if (oDto == null)
                    oDto = ActiveItems.Find(x => oBrand.Equals(x.Product));
            }
            else if (oProduct is DTOProductBrand oBrand)
            {
                oDto = ActiveItems.Find(x => oBrand.Equals(x.Product));
            }

            if (oDto == null)
                oDto = ActiveItems.Find(x => x.Product == null);
            if (oDto != null)
                retval = oDto.Dto;
            return retval;
        }
    }
}

File: matiasmasso/Mat.NET/DT2/DTOCustomerTarifaDto.vb
Public Class DTOCustomerTarifaDto
    Inherits DTOBaseGuid

    Property CustomerOrChannel As DTOBaseGuid
    Property CustomerOrChannelNom As String
    Property Fch As Date
    Property Product As DTOProduct
    Property Dto As Decimal
    Property Src As Srcs
    Property Obs As String

    Public Enum Srcs
        NotSet
        Canal
        Client
    End Enum

    Public Sub New()
        MyBase.New
    End Sub


    Public Sub New(oGuid As Guid)
        MyBase.New(oGuid)
    End Sub

    Shared Function Factory(oCustomerOrChannel As DTOBaseGuid, Optional Nom As String = "")
        Dim retval As New DTOCustomerTarifaDto
        With retval
            .CustomerOrChannel = oCustomerOrChannel
            If TypeOf oCustomerOrChannel Is DTOContact Then
                .Src = Srcs.Client
            ElseIf TypeOf oCustomerOrChannel Is DTODistributionChannel Then
                .Src = Srcs.Canal
            End If
            .CustomerOrChannelNom = Nom
            .Fch = Now
        End With
        Return retval
    End Function

    Shared Function FullNom(oCustomerDto As DTOCustomerTarifaDto, Optional oLang As DTOLang = Nothing) As String
        If oLang Is Nothing Then oLang = DTOApp.current.lang
        Dim sAllBrands As String = oLang.tradueix("todas las marcas", "totes les marques", "all brands")
        Dim sConcept As String = IIf(oCustomerDto Is Nothing, sAllBrands, oCustomerDto.Product.Nom)
        Dim retval As String = String.Format("{0:dd/MM/yy} {1}", oCustomerDto.Fch, sConcept)
        Return retval
    End Function

    Shared Function ProductDto(ActiveItems As List(Of DTOCustomerTarifaDto), oProduct As DTOProduct) As Decimal
        Dim retval As Decimal
        Dim oSku As DTOProductSku = Nothing
        Dim oCategory As DTOProductCategory = Nothing
        Dim oBrand As DTOProductBrand = Nothing

        Dim oDto As DTOCustomerTarifaDto = Nothing
        If TypeOf oProduct Is DTOProductSku Then
            oSku = oProduct
            oCategory = oSku.Category
            oBrand = oCategory.Brand
            oDto = ActiveItems.Find(Function(x) oSku.Equals(x.Product))
            If oDto Is Nothing Then
                oDto = ActiveItems.Find(Function(x) oCategory.Equals(x.Product))
                If oDto Is Nothing Then
                    oDto = ActiveItems.Find(Function(x) oBrand.Equals(x.Product))
                End If
            End If
        ElseIf TypeOf oProduct Is DTOProductCategory Then
            oCategory = oProduct
            oBrand = oCategory.Brand
            oDto = ActiveItems.Find(Function(x) oCategory.Equals(x.Product))
            If oDto Is Nothing Then
                oDto = ActiveItems.Find(Function(x) oBrand.Equals(x.Product))
            End If
        ElseIf TypeOf oProduct Is DTOProductBrand Then
            oBrand = oProduct
            oDto = ActiveItems.Find(Function(x) oBrand.Equals(x.Product))
        End If

        If oDto Is Nothing Then
            oDto = ActiveItems.Find(Function(x) x.Product Is Nothing)
        End If
        If oDto IsNot Nothing Then
            retval = oDto.Dto
        End If
        Return retval
    End Function
End Class
